internal/controllers/iplease_controller.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2021 The routerd authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	goipam "github.com/metal-stack/go-ipam"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ipamv1alpha1 "routerd.net/kube-ipam/api/v1alpha1"
+	"routerd.net/kube-ipam/internal/controllers/adapter"
+)
+
+// IPLeaseReconciler reconciles a IPLease object
+type IPLeaseReconciler struct {
+	client.Client
+	Log       logr.Logger
+	Scheme    *runtime.Scheme
+	IPAMCache ipamCache
+
+	IPPoolType  adapter.IPPool
+	IPLeaseType adapter.IPLease
+}
+
+// +kubebuilder:rbac:groups=ipam.routerd.net,resources=ipleases,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ipam.routerd.net,resources=ipleases/status,verbs=get;update;patch
+
+func (r *IPLeaseReconciler) Reconcile(
+	ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	log := r.Log.WithValues("iplease", req.NamespacedName)
+
+	iplease := adapter.AdaptIPLease(
+		r.IPLeaseType.ClientObject().DeepCopyObject())
+	if err = r.Get(ctx, req.NamespacedName, iplease.ClientObject()); err != nil {
+		return res, client.IgnoreNotFound(err)
+	}
+	defer func() {
+		// ensure that no matter how we exit the reconcile function,
+		// we want to reconcile the IPLease after the lease duration expired.
```

Comment:
Is it really needed to wait for lease expiration when a lease already has been deleted?